src/internet_identity/src/archive.rs

```diff
@@ -0,0 +1,179 @@
+use crate::state;
+use crate::state::{ArchiveData, ArchiveState, DeviceDataInternal};
+use candid::Principal;
+use ic_cdk::api::call::CallResult;
+use ic_cdk::api::management_canister::main::{
+    canister_status, create_canister, install_code, CanisterIdRecord, CanisterInstallMode,
+    CanisterInstallMode::Install, CreateCanisterArgument, InstallCodeArgument,
+};
+use ic_cdk::api::time;
+use ic_cdk::{id, notify};
+use internet_identity_interface::{
+    ArchiveInit, DeployArchiveResult,
+    DeployArchiveResult::{CreationFailed, UpgradeFailed},
+    DeviceDataUpdate, Entry, Operation, Private, UserNumber,
+};
+use serde_bytes::ByteBuf;
+use sha2::Digest;
+use sha2::Sha256;
+use CanisterInstallMode::Upgrade;
+
+pub async fn deploy_archive(wasm: ByteBuf) -> DeployArchiveResult {
+    let archive_state = state::persistent_state_mut(|persistent_state| {
+        let archive_state = persistent_state.archive_info.state.clone();
+        if archive_state == ArchiveState::NotCreated {
+            // lock archive creation because of async operation
+            persistent_state.archive_info.state = ArchiveState::CreationInProgress;
+        }
+        archive_state
+    });
+
+    match archive_state {
+        ArchiveState::NotCreated => create_and_install_archive(wasm.into_vec()).await,
+        ArchiveState::CreationInProgress => DeployArchiveResult::CreationInProgress,
+        ArchiveState::Created(data) => {
+            match install_archive(data.archive_canister, wasm.into_vec()).await {
+                Ok(()) => DeployArchiveResult::Success(data.archive_canister),
+                Err(err) => UpgradeFailed(err),
+            }
+        }
+    }
+}
+
+async fn create_and_install_archive(wasm: Vec<u8>) -> DeployArchiveResult {
+    match create_archive().await {
+        Err(e) => {
+            state::persistent_state_mut(|persistent_state| {
+                // unlock archive creation again
+                persistent_state.archive_info.state = ArchiveState::NotCreated
```

Comment:
I'm a bit uneasy about the control flow here; if for any reason (bug, canister trapping, etc) we fail before this line then we'll forever be stuck in `CreationInProgress`, right? Also, having the state being set somewhere (in `deploy_archive` and reverted elsewhere (here) is not particularly clear and may lead to some nasty bugs down the road

Reply:
Hm, I agree. Let me reorganize things to make it clearer. Also, there is an inherent risk of trapping during canister creation because we need to send a message which might fail for a whole host of reasons. So I'll add a timestamp to the `CreationInProgress` lock, which will unlock new archive canister creation after some reasonably long time (24 hours?).